cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/DefaultHttpServiceResponder.java

```diff
@@ -154,10 +152,20 @@ public void sendError(int status, String errorMessage) {
    * @param headers the headers to be sent back
    */
   @Override
-  public void send(int status, ByteBuffer content, String contentType, Multimap<String, String> headers) {
-    responder.sendContent(HttpResponseStatus.valueOf(status),
-                          ChannelBuffers.wrappedBuffer(content), contentType, headers);
-    emitMetrics(status);
+  public void send(final int status, ByteBuffer content, final String contentType, Multimap<String, String> headers) {
+    bufferedResponse = new BufferedResponse(status, content, contentType, headers);
```

Comment:
Do you want to protect against sending twice? The responder would throw an exception if you send two responses ("response has already been sent"). That error would go unnoticed now. I think it is better if we log at least a warning for that, if not throw an exception so that developers are aware of their buggy handlers.